File: EJERCICIOS/SEGUNDO PARCIAL/CONSULTAS/Consulta01/O_Radix/radix.cpp
#include <iostream>
#include <vector>
#include <algorithm>

void radixSort(std::vector<int>& arr) {
    int maxVal = *std::max_element(arr.begin(), arr.end());

    for (int exp = 1; maxVal / exp > 0; exp *= 10) {
        countingSort(arr, exp);
    }
}

void countingSort(std::vector<int>& arr, int exp) {
    int base = 10; // Base decimal

    int n = arr.size();
    std::vector<int> count(base, 0); // Inicializar contador

    // Contar la frecuencia de los dígitos
    for (int num : arr) {
        int digit = (num / exp) % base;
        count[digit]++;
    }

    // Calcular las posiciones finales de los dígitos
    for (int i = 1; i < base; i++) {
        count[i] += count[i - 1];
    }

    // Construir el arreglo ordenado
    std::vector<int> sortedArr(n);
    for (int i = n - 1; i >= 0; i--) {
        int num = arr[i];
        int digit = (num / exp) % base;
        sortedArr[count[digit] - 1] = num;
        count[digit]--;
    }

    // Actualizar el arreglo original con el arreglo ordenado
    arr = sortedArr;
}

int main() {
    std::vector<int> arr = {170, 45, 75, 90, 802, 24, 2, 66};
    radixSort(arr);
    for (int num : arr) {
        std::cout << num << " ";
    }
    return 0;
}

File: EJERCICIOS/SEGUNDO PARCIAL/CONSULTAS/Consulta01/O_Radix/radix.py
def radixSort(arr):
    # Encontrar el máximo valor en la lista
    max_val = max(arr)

    # Realizar el ordenamiento por cada dígito
    exp = 1
    while max_val // exp > 0:
        countingSort(arr, exp)
        exp *= 10

def countingSort(arr, exp):
    base = 10  # Base decimal

    n = len(arr)
    count = [0] * base  # Inicializar contador

    # Contar la frecuencia de los dígitos
    for num in arr:
        digit = (num // exp) % base
        count[digit] += 1

    # Calcular las posiciones finales de los dígitos
    for i in range(1, base):
        count[i] += count[i - 1]

    # Construir el arreglo ordenado
    sorted_arr = [0] * n
    for num in reversed(arr):
        digit = (num // exp) % base
        sorted_arr[count[digit] - 1] = num
        count[digit] -= 1

    # Actualizar la lista original con el arreglo ordenado
    for i in range(n):
        arr[i] = sorted_arr[i]

# Ejemplo de uso
arr = [170, 45, 75, 90, 802, 24, 2, 66]
radixSort(arr)
print(arr)
